### ISSUE TITLE
Validation on test dataset gives zero metrics

### ISSUE THREAD

**Issue by pheersaquoff** [ultralytics/ultralytics#3303]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I try to use val method for testing my model, I pass yaml file with dataset different from training which looks like this 
```
-images
---val
-labels
---val
```
with no 'train' directories. val metrics are obviously too small and predictions in ```val_batch{I}_labels.jpg``` are noticeably different from predictions after predict() method on ```images/val``` directory with same conf and iou

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@pheersaquoff likely user error for incorrect dataset formatting. See https://docs.ultralytics.com/datasets/detect/

**Comment by pheersaquoff** [comment]
@glenn-jocher i made exact same formatting and added train images and labels to corresponding directories. so i trained with one yaml and validate with another, train parts are the same but val parts are different. and i still get the same

**Comment by pheersaquoff** [comment]
@glenn-jocher any thoughts please?

**Comment by glenn-jocher** [comment]
@pheersaquoff certainly, it appears that the issue might be how you formatted your validation dataset. You mentioned that you trained on one dataset and validated using a different dataset, which is fine. However, did you ensure that the validation dataset has the same formatting as the training set? Specifically, did you have the 'train' and 'val' directories in the same locations and structure in both datasets? 

If your datasets have the same structure and formatting and you are still getting zero metrics in validation, you may want to check if your validation annotations are correct and match the ground truth. A good way to check this is to visualize the validation set predictions and compare them with the ground truth. 

Lastly, if the above checks do not fix the issue, please provide more information and a minimal reproducible example, so we can take a deeper look.

**Comment by pheersaquoff** [comment]
@glenn-jocher yes i checked all of this, structure is correct
```
-images
---train
---val
-labels
---train
---val
```
and ```train``` direcroties are empty.
detect method is working good on val image with same confidence & iou.
what can be a minimal reproducible example in this case?
Also my dataset is single class

**Comment by glenn-jocher** [comment]
@pheersaquoff hello,

Thank you for providing the information. It seems that the dataset structure and formatting are correct as you have verified. Since your dataset is single class, the issue might not be related to the number of classes.

To proceed with debugging, we can start by verifying the annotations in your validation dataset. You can visually compare the predictions made using the `detect` method on a validation image with the corresponding ground truth labels. This will help ensure that the annotations are correct and aligned properly.

If the annotations are accurate and aligned, we can then investigate other potential factors. You mentioned that the `val` metrics are significantly different from the predictions made using the `detect` method. In this case, it might be helpful to examine the specific differences between the two. For example, are the bounding box coordinates or predicted classes different?

Additionally, providing a minimal reproducible example, such as a specific image and its corresponding annotations, may help in further understanding the issue and finding a solution.

Please share any relevant findings or additional information so we can continue to assist you.

**Comment by pheersaquoff** [comment]
@glenn-jocher I double checked everything and figured out that val() method uses different confidence score to compute metrics from what I specify in arguments. so minimal reproducible example is
```
from ultralytics import YOLO
model = YOLO('yolov8l.pt')  # load an official model
model = YOLO('path/to/my.pt')  # load a custom model
metrics = model.val(
    data = 'dataset.yaml',
    conf=0.001
    )
```

but it prints metrics with another conf as I see in recall-confidence curve

**Comment by glenn-jocher** [comment]
@pheersaquoff hi there,

Thank you for bringing this to our attention. I understand that you are experiencing an issue with the `val()` method in YOLOv8, where the confidence score used to compute metrics seems to be different from the one specified in the arguments.

To investigate this issue further, I suggest checking if there are any default values or configurations that might affect the confidence score used in the `val()` method. Additionally, you can inspect the code related to the `val()` method to understand how the confidence score is being used and if there are any transformations or adjustments applied to it.

Furthermore, it might be helpful to compare the metrics computed by the `val()` method with the recall-confidence curve to validate if they align properly. This can provide insights into whether the specified confidence score is being used correctly in the metric calculations.

Please let us know if you have any additional information or findings related to this issue. We'll do our best to assist you further.

**Comment by SBhat2615** [comment]
@glenn-jocher Hello.
I'm using roboflow colab notebook for object detection, which uses CLI commands. I see only validation set printing the metrics like precision, recall & mAP. How can i get those values for test set as well??

**Comment by SBhat2615** [comment]
Should i not use test data seperately then ??
Also, split=test isn't working for yolov8..

There's one for validation 
!yolo task=detect mode=val model={HOME}/runs/detect/train/weights/best.pt data={dataset.location}/data.yaml

And one for test
!yolo task=detect mode=predict model={HOME}/runs/detect/train/weights/best.pt conf=0.25 source={dataset.location}/test/images save=True

**Comment by pderrenger** [comment]
@SBhat2615 in YOLOv8, the concept of validation and test sets typically serves different purposes. While validation is used during training to evaluate the model's performance and make decisions about training, the test set is used to assess the model's generalization to new, unseen data.

To obtain metrics for the test set, you can utilize the following command:
```
!yolo task=detect mode=val model={HOME}/runs/detect/train/weights/best.pt data={dataset.location}/data.yaml split=test
```

This command specifically instructs YOLOv8 to perform validation using the test set and should provide you with the desired metrics for the test set.

Regarding your point about the validation and test sets being the same in YOLOv8, they are indeed separate datasets with distinct purposes. While the model undergoes validation during training, the test set is crucial for assessing the model's performance on unseen data and ensuring its generalization capabilities.

Lastly, you mentioned the lack of metrics for predicted images in the second command. If you wish to obtain metrics for the predicted images, you'd typically need ground truth annotations for those predicted images in order to evaluate the model's predictions against the actual objects present in the images. This is often done during the evaluation phase using separate scripts or tools for evaluation and comparison against ground truth annotations.

Feel free to clarify any additional questions or concerns you may have.

**Comment by SBhat2615** [comment]
> @SBhat2615 in YOLOv8, the concept of validation and test sets typically serves different purposes. While validation is used during training to evaluate the model's performance and make decisions about training, the test set is used to assess the model's generalization to new, unseen data.
> 
> To obtain metrics for the test set, you can utilize the following command:
> 
> ```
> !yolo task=detect mode=val model={HOME}/runs/detect/train/weights/best.pt data={dataset.location}/data.yaml split=test
> ```
> 
> This command specifically instructs YOLOv8 to perform validation using the test set and should provide you with the desired metrics for the test set.
> 
> Regarding your point about the validation and test sets being the same in YOLOv8, they are indeed separate datasets with distinct purposes. While the model undergoes validation during training, the test set is crucial for assessing the model's performance on unseen data and ensuring its generalization capabilities.
> 
> Lastly, you mentioned the lack of metrics for predicted images in the second command. If you wish to obtain metrics for the predicted images, you'd typically need ground truth annotations for those predicted images in order to evaluate the model's predictions against the actual objects present in the images. This is often done during the evaluation phase using separate scripts or tools for evaluation and comparison against ground truth annotations.
> 
> Feel free to clarify any additional questions or concerns you may have.

Thank you for the reply.
Actually i'm using roboflow for annotation & it gives the data in this format. Below attached is yaml file.

'0'
nc: 1
roboflow:
license: CC BY 4.0
project: detection-dljca
url: https://universe.roboflow.com/detection-twelqh/detection-dwejca/dataset/3
version: 3
workspace: detection-twelqh
test: ../test/images

train: Detection-3/train/images
val: Detection-3/valid/images

I was trying to get metrics along with the prediction for the test set. I do have the ground truth, not sure which command to use as mode=predict gave the prediction but without any metrics.

This is the directory structure.
<img width="220" alt="Screenshot 2023-12-14 at 5 30 12 AM" src="https://github.com/ultralytics/ultralytics/assets/77118437/2897ff82-871e-43d3-946e-84d24b78f289">

**Comment by glenn-jocher** [comment]
@SBhat2615 to get test metrics in YOLOv8, you need to ensure your dataset YAML includes a test set path. Given your provided YAML structure, modify the `test` entry to include both images and labels directories, like this:

```yaml
test:
  images: ../test/images
  labels: ../test/labels  # This line should point to your test labels
```

Then, run the val command with `data` and `model` set to your paths, which will now calculate metrics on your test set given that it has ground truth annotations:

```
!yolo task=detect mode=val data={your_dataset_yaml} model={your_best_weights.pth}
```

Make sure `{your_dataset_yaml}` points to the path of your modified YAML file, and `{your_best_weights.pth}` to your model's weights. This should output the desired metrics for your test set.

**Comment by SBhat2615** [comment]
> @SBhat2615 to get test metrics in YOLOv8, you need to ensure your dataset YAML includes a test set path. Given your provided YAML structure, modify the `test` entry to include both images and labels directories, like this:
> 
> ```yaml
> test:
>   images: ../test/images
>   labels: ../test/labels  # This line should point to your test labels
> ```
> 
> Then, run the val command with `data` and `model` set to your paths, which will now calculate metrics on your test set given that it has ground truth annotations:
> 
> ```
> !yolo task=detect mode=val data={your_dataset_yaml} model={your_best_weights.pth}
> ```
> 
> Make sure `{your_dataset_yaml}` points to the path of your modified YAML file, and `{your_best_weights.pth}` to your model's weights. This should output the desired metrics for your test set.

It again gave metrics for validation set for the below command.

!yolo task=detect mode=val model={HOME}/runs/detect/train/weights/best.pt data={dataset.location}/data.yaml save_json=True

test:
  images: ../test/images
  labels: ../test/labels
train: Fungus-Detection-3/train/images
val: Fungus-Detection-3/valid/images

If i do mode=val how will it know to give results for the test set?
I've tried split=test, but doesn't work for v8 as i said in other thread.

Thanks for quick replies!!!

**Comment by glenn-jocher** [comment]
@SBhat2615 to calculate metrics on your test set with YOLOv8, ensure your dataset YAML correctly references your test images and labels. Then, use the `val` command specifying your dataset and model:

```
!yolo task=detect mode=val data={dataset.location}/data.yaml model={HOME}/runs/detect/train/weights/best.pt
```

This command will automatically use the test set for metrics if the `val` split is empty and a `test` split is specified in your YAML. If issues persist, ensure the `test` path in your YAML is properly formatted and pointing to the correct directories with your test data and annotations.
